### PR TITLE
Special characters in object name in list objects

### LINKED ISSUE #943: listObjectsV2 when there is & in object name
Version: **7.0.18**
Operation: **listObjectsV2**

When calling putObject operation for a file that contains the char & (example&.pdf), the file is correctly uploaded on minio. The created object has the name example&.pdf. If this object 's key is test/example&.pdf , when we call listObjectsV2 with prefix 'test/' the object that is returned, has the name example`&amp;`.pdf instead of the correct example&.pdf

**It is working correctly for version 7.0.16**

### PR DESCRIPTION
**Pull request by prakashsvmx**
Decode special characters in object name in list object methods:
`listObjects`
`listObjectsV2`
`extensions.listObjectsV2WithMetadata`

Fixes #943

Tested Against
- [x] MinIO Server
- [x] S3

<summary> Test JS Code  </summary>

<details>

```


var Minio = require("../../dist/main/minio")
const _ = require('lodash')

// Instantiate the minio client with the endpoint
// and access keys as shown below.
//Play instance
var minioClient = new Minio.Client({
  endPoint: 'play.min.io',
  port: 9000,
  useSSL: true,
  accessKey: 'Q3AM3UQ867SPQQA43P2F',
  secretKey: 'zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG'
})

//Local testing
minioClient = new Minio.Client({
  endPoint: 'localhost',
  port: 9000,
  useSSL: false,
  accessKey: 'minio',
  secretKey: 'minio123'
})

const file = '/home/user/Downloads/Temp/MinIOJSTestObjects/example&.pdf'
const bucketName = "sph-my-bucket"
const objectName="examp le&0a!-_.*'()&$@=;:+,?<>.pdf"
const prefix="test"
const objectNameWithPrefix = `${prefix}/${objectName}`

function fputObjectTopLevel () {

  var metaData = {
    'Content-Type': 'application/octet-stream',
    'X-Amz-Meta-File-Name': objectName
  }
  // Using fPutObject API upload your file to the bucket europetrip.

  minioClient.fPutObject(bucketName, objectName , file, metaData, function (err, info) {
    if (err) return console.log(err)
    console.log('File Fput uploaded successfully.',info)
  })
}

function fputObjectPrefixLevel () {


  var metaData = {
    'Content-Type': 'application/octet-stream',
    'X-Amz-Meta-File-Name': objectNameWithPrefix
  }
  // Using fPutObject API upload your file to the bucket europetrip.

  minioClient.fPutObject(bucketName, objectNameWithPrefix , file, metaData, function (err, info) {
    if (err) return console.log(err)
    console.log('File Fput uploaded successfully.',info)
  })
}


const listObjectsByFnVersion =(listApiVersion, cb)=>{
  const bucketName = "sph-my-bucket"
  var objectsStream
  if(listApiVersion===1){
    objectsStream=minioClient.listObjects(bucketName, "", true )
  }else if(listApiVersion ===2) {
    objectsStream = minioClient.listObjectsV2(bucketName, "", true)
  }else if(listApiVersion ===3) {
    objectsStream = minioClient.extensions.listObjectsV2WithMetadata(bucketName, "",true)
  }
  objectsStream.on('data', function (obj) {

    //console.log("Entry:: ::",listApiVersion, obj.prefix, obj.name, obj.versionId)

    cb && cb()
  })
  objectsStream.on('end',()=>{
    cb && cb()
  })
  objectsStream.on('error', function (e) {
    console.log("::Error:",e)
  })
}

const getObjectWithoutPrefix = () =>{
    
  minioClient.getObject(bucketName, objectName , (e, dataStream) => {
    if (e) return e
    dataStream.on('data', function(chunk) {
      console.log("Data. Total size = " ,dataStream.headers)
    })
    dataStream.on('end', function() {
      console.log("End. Total size = " ,dataStream.headers)
    })
    dataStream.on('error', function(e) {
      console.log(e)
    })
  })

    
}

const getObjectWithPrefix = () =>{

  minioClient.fGetObject(bucketName, objectNameWithPrefix,`./test/${objectNameWithPrefix}`).then((err, res)=>{
    if(!err) {
      console.log("Got Object", objectNameWithPrefix)
    }
  })
}

const deleteObject =()=>{
  //Delete Top level
  minioClient.removeObject(bucketName, objectName).then(( e)=>{
    if(!e) {
      console.log("Done Delete...", objectName)
    }else {
      console.log("Failed to delete")
    }
  })

  //Delete second level.
    
  minioClient.removeObject(bucketName, objectNameWithPrefix).then((e)=>{
    if(!e) {
      console.log("Done Delete..." ,objectNameWithPrefix)
    }else {
      console.log("Failed to delete")
    }
  })
    
}

const statObjectTopLevel = () =>{
  minioClient.statObject(bucketName , objectName).then((e,res)=>{
    console.log(":Stat on:",objectName, e, res)
  })
}

const statObjectPrefixLevel = () =>{
  minioClient.statObject(bucketName , objectNameWithPrefix).then((e,res)=>{
    console.log("Stat on:",objectNameWithPrefix, e, res)
  })
}


//fputObjectTopLevel()
//fputObjectPrefixLevel()
//listObjects(1)
//listObjects(2)
//listObjects(3)
//getObjectWithoutPrefix()
//getObjectWithPrefix()
//statObjectTopLevel()
//statObjectPrefixLevel()

//deleteObject()

```
</details>

### CODE REVIEW

**Review by kannappanr — APPROVED** [review]
LGTM

**Review by kannappanr — APPROVED** [review]
LGTM

**Review by kannappanr — DISMISSED** [review]
LGTM

File: src/test/functional/functional-tests.js [hunk 1]
```diff
@@ -2512,4 +2512,210 @@ describe('functional tests', function() {
...
+
+      step(`putObject(bucketName, objectName, stream)_bucketName:${bucketNameForSpCharObjects}, _objectName:${objectNameSpecialChars}, stream:100Kib_`, done => {
+        client.putObject(bucketNameForSpCharObjects, objectNameSpecialChars, objectContents)
+          .then(() => {
+            done()
+          })
+          .catch(done)
+      })
+
+      step(`listObjects(bucketName, prefix, recursive)_bucketName:${bucketNameForSpCharObjects}, prefix:"/, true,`, done => {
```
Comment (harshavardhana) [inline src/test/functional/functional-tests.js:2538]:
```suggestion
      step(`listObjects(bucketName, prefix, recursive)_bucketName:${bucketNameForSpCharObjects}, prefix:"", true,`, done => {
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: src/test/functional/functional-tests.js (hunk 1 above)
Comment (prakashsvmx) [reply src/test/functional/functional-tests.js]: Thank you @harshavardhana . I have cleaned it up.

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
Some conflicts need to be resolved.

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
PTAL @kannappanr

**Comment by harshavardhana** [comment]
Taking this in.

**Pull request merged**